src/main/java/org/folio/circulation/resources/RequestByInstanceIdResource.java

```diff
@@ -206,16 +229,90 @@ private void createInstanceLevelRequests(RoutingContext routingContext) {
     return succeeded(requests);
   }
 
-  private Result<Collection<Item>> getAvailableItems(Collection<Item> items) {
+  private Result<InstanceRequestRelatedRecords> segregateItemsList(Collection<Item> items,
+                                                                          InstanceRequestRelatedRecords requestRelatedRecords ){
     if (items == null ||items.isEmpty()) {
       return failedValidation("Items list is null or empty", "items", "null");
     }
 
-    return succeeded(items.stream()
-      .filter(Item::isAvailable)
-      .collect(Collectors.toList()));
+    requestRelatedRecords.setUnsortedAvailableItems(
+      items.stream()
+        .filter(Item::isAvailable)
+        .collect(Collectors.toList()));
+
+    requestRelatedRecords.setUnsortedUnavailableItems(
+      items.stream()
+        .filter(item -> !item.isAvailable())
+        .collect(Collectors.toList()));
+
+    return succeeded(requestRelatedRecords);
   }
 
+  private static CompletableFuture<Result<InstanceRequestRelatedRecords>> combineWithUnavailableItems(InstanceRequestRelatedRecords records, Clients clients){
+
+    RequestQueueRepository queueRepository = RequestQueueRepository.using(clients);
+
+    final Collection<Item> unsortedUnavailableItems = records.getUnsortedUnavailableItems();
+
+    Map<Item, CompletableFuture<Result<RequestQueue>>> itemRequestQueueMap = new HashMap<>();
+    if (unsortedUnavailableItems == null || unsortedUnavailableItems.isEmpty()) {
+      return CompletableFuture.completedFuture(succeeded(records));
+    }
+
+    for (Item item : unsortedUnavailableItems) {
+      itemRequestQueueMap.put(item, queueRepository.getLiteRequestQueues(item.getItemId()));
+    }
+
+    final Collection<CompletableFuture<Result<RequestQueue>>> requestQueueFutures = itemRequestQueueMap.values();
+
+    //Collect the RequestQueue objects once they come back
+    return CompletableFuture.allOf(requestQueueFutures.toArray(new CompletableFuture[requestQueueFutures.size()]))
+      .thenApply(x -> {
+        Map<Item, RequestQueue> itemQueueSizeMap = new HashMap<>();
+        for (Map.Entry<Item, CompletableFuture<Result<RequestQueue>>> entry : itemRequestQueueMap.entrySet()) {
+          Result<RequestQueue> requestQueueResult = entry.getValue().join();
+          if (requestQueueResult.succeeded()) {
+            itemQueueSizeMap.put(entry.getKey(), requestQueueResult.value());
+          }
+        }
+
+        //Sort the map
+        Map<Item, RequestQueue> sortedMap = sortRequestQueues(itemQueueSizeMap);
```

Comment:
Although it's more streamlined to return the keyset (a list of items), I would like to keep the method's signature this way as it more clearly defines the separation of responsibilities: an unsorted map in, sorted map out, in case some theoretical caller wants to use the method differently, and it makes unit testing it easier. I'll add more tests for this method.

Reply:
Are these maps used as maps, or are they just a handy way of keeping a pair of objects together?

Reply:
One aspect of using HashMaps is that they do not guarantee the [order of iteration](https://docs.oracle.com/javase/8/docs/api/java/util/HashMap.html)

Reply:
If we want to retain the map, another way to achieve this is:
```
    //Use linked hash map to guarantee sort order
    return unsortedItems
      .entrySet()
      .stream()
      .sorted(compareQueueLengths())
      .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue,
        (oldValue, newValue) -> oldValue, (LinkedHashMap::new)));
```

where `compareQueueLengths()` is the comparison extracted to a method.

```
  private static Comparator<Map.Entry<Item, RequestQueue>> compareQueueLengths() {
    return (q1, q2) -> {
      RequestQueue queue1 = q1.getValue();
      RequestQueue queue2 = q2.getValue();

      int result = queue1.size() - queue2.size();

      if (result == 0) {
        return queue1.getLowestPriorityFulfillableRequest()
          .getRequestExpirationDate()
          .compareTo(queue2.getLowestPriorityFulfillableRequest().getRequestExpirationDate());
      }

      return result;
    };
  }
```

Reply:
Yes, the use of hashmap here is to keep a pair of objects together with guaranteeing order.

Reply:
Do we even need a map? Doesn't each `Request` in each RequestQueue, assuming there are any, have an `itemId` that we could use? Maybe something we can do in the future to simplify this.

Reply:
I think we need to have a map, because as you said, the RequestQueue may not have a Request. I added a check for it last night and adding tests for this case now.

Reply:
My thought was for any item id in the sorted list of un/available items, if there is not a request queue that has that item id, then there are no requests for that item, so the length is 0. But I don't want to belabor this, since I have no idea if what I am saying is useful or would work. Since you are confident this already works, we can table this for a future discussion or just ignore it since we know this solution is not the ideal one. 😄